### DISCUSSION TITLE
Relative path for test data

### DISCUSSION THREAD

**Discussion by blackliner** [bazelbuild/bazel#19558] [General]
Is it possible to reference my `data.json` in a "package local" way, instead of having to put in the whole absolute path from the WORKSPACE root? Having the full path makes it difficult to use the same code that runs in the bazel workspace vs final deployments.

Is there maybe a way to tell bazel to copy the data files right next to the binaries?

**Comment by meteorcloudy** [comment]
Sounds like you are looking for https://bazel.build/extending/rules#runfiles, basically you can specify your data dependencies through the `data` attribute and they should be available during runtime via its runfiles path (package related).

It's also recommended to use the runfiles libraries to access them:

- C++: [`@bazel_tools//tools/cpp/runfiles`](https://github.com/bazelbuild/bazel/blob/7ff98ba28a21025fccdf82a3d148407be48fae40/tools/cpp/runfiles/runfiles_src.h#L17)
- Java: [`@bazel_tools//tools/java/runfiles`](https://github.com/bazelbuild/bazel/blob/7ff98ba28a21025fccdf82a3d148407be48fae40/tools/java/runfiles/Runfiles.java#L33)
- Bash: [`@bazel_tools//tools/bash/runfiles`](https://github.com/bazelbuild/bazel/blob/7ff98ba28a21025fccdf82a3d148407be48fae40/tools/bash/runfiles/runfiles.bash#L55)
- Python: [bazel-runfiles library](https://github.com/bazelbuild/rules_python/blob/main/python/runfiles/README.rst#bazel-runfiles-library)
